alf/layers.py

```diff
@@ -733,6 +733,159 @@ def bias(self):
+
+        Args:
+            input_size (int): input size
+            output_size (int): output size
+            n (int): the size of the paramster set
+            activation (torch.nn.functional):
+            use_bn (bool): whether use Batch Normalization.
+            use_ln (bool): whether use layer normalization
+            use_bias (bool): whether use bias
+            kernel_initializer (Callable): initializer for the FC layer kernel.
+                If none is provided a ``variance_scaling_initializer`` with gain
+                as ``kernel_init_gain`` will be used.
+            kernel_init_gain (float): a scaling factor (gain) applied to
+                the std of kernel init distribution. It will be ignored if
+                ``kernel_initializer`` is not None.
+            bias_init_value (float): a constant
+        """
+        super().__init__()
+        self._activation = activation
+        self._weight = nn.Parameter(torch.Tensor(n, output_size, input_size))
+        if use_bias:
+            self._bias = nn.Parameter(torch.Tensor(n, output_size))
+        else:
+            self._bias = None
+
+        self._kernel_initializer = kernel_initializer
+        self._kernel_init_gain = kernel_init_gain
+        self._bias_init_value = bias_init_value
+        self._use_bias = use_bias
+        self._use_bn = use_bn
+        self._use_ln = use_ln
+        self._n = n
+
+        if use_bn:
+            self._bn = nn.BatchNorm1d(output_size)
+        else:
+            self._bn = None
+        if use_ln:
+            self._ln = nn.LayerNorm(output_size)
+        else:
+            self._ln = None
+        self.reset_parameters()
+
+    def forward(self, inputs, comp_weight=None):
+        """Forward
+
+        Args:
+            inputs (torch.Tensor): with shape ``[B, input_size]``
+            comp_weight (torch.Tensor | None): with shape ``[B, n]``.
+                Uniform weight of one wil be used if None.
+        Returns:
+            torch.Tensor with shape ``[B, output_size]``
+        """
+        n, k, l = self._weight.shape
+        if inputs.ndim == 2:
+            assert inputs.shape[1] == l, (
+                "inputs has wrong shape %s. Expecting (B, %d)" % (inputs.shape,
+                                                                  l))
+            inputs = inputs.unsqueeze(0).expand(n, *inputs.shape)
+
+        else:
+            raise ValueError("Wrong inputs.ndim=%d" % inputs.ndim)
+
+        if self.bias is not None:
+            y = torch.baddbmm(
+                self._bias.unsqueeze(1), inputs,
+                self.weight.transpose(1, 2))  # [n, B, k]
+        else:
+            y = torch.bmm(inputs, self._weight.transpose(1, 2))  # [n, B, k]
+        y = y.transpose(0, 1)  # [B, n, k]
+
+        if comp_weight is not None:
+            assert comp_weight.ndim == 2, (
+                "Wrong comp_weight.ndim=%d" % comp_weight.ndim)
+
+            y = y * comp_weight.unsqueeze(-1)
```

Comment:
you can save the sum at line 835 by using matmul
```python
y = torch.matmul(comp_weight.unsqueeze(1), y).squeeze(1)
```

Reply:
Changed to use ``torch.bmm`` to avoid the additional ``sum``.